jsonrpc/server.go

```diff
@@ -169,59 +172,81 @@ func (s *Server) HandleReader(reader io.Reader) ([]byte, error) {
 		}
 	} else {
 		var batchReq []json.RawMessage
-		var batchRes []json.RawMessage
 
 		if batchJSONErr := dec.Decode(&batchReq); batchJSONErr != nil {
 			res.Error = Err(InvalidJSON, batchJSONErr.Error())
 		} else if len(batchReq) == 0 {
 			res.Error = Err(InvalidRequest, "empty batch")
 		} else {
-			for _, rawReq := range batchReq { // todo: handle async
-				var resObject *response
+			return s.handleBatchRequest(ctx, batchReq)
+		}
+	}
+
+	if res == nil {
+		return nil, nil
+	}
+	return json.Marshal(res)
+}
+
+func (s *Server) handleBatchRequest(ctx context.Context, batchReq []json.RawMessage) ([]byte, error) {
+	var batchRes []json.RawMessage
+	var resMutex sync.Mutex
+
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	p := pool.New().WithErrors()
+	for _, rawReq := range batchReq {
+		rawReq := rawReq
+		p.Go(func() error {
+			var resObject *response
 
-				reqDec := json.NewDecoder(bytes.NewBuffer(rawReq))
-				reqDec.UseNumber()
+			reqDec := json.NewDecoder(bytes.NewBuffer(rawReq))
+			reqDec.UseNumber()
 
-				req := new(request)
-				if jsonErr := reqDec.Decode(req); jsonErr != nil {
+			req := new(request)
+			if jsonErr := reqDec.Decode(req); jsonErr != nil {
+				cancel()
```

Comment:
requests should are handled separately. so even if one of them is not a valid request object, the rest should still be processed.

only exception is if ```		if batchJSONErr := dec.Decode(&batchReq); batchJSONErr != nil {
			res.Error = Err(InvalidJSON, batchJSONErr.Error())``` fails. In that case, we fail the entire batch with invalid json error

Reply:
Removed `cancel()`